Add `Comments` section in TASTY

Duhemm opened the pull request:
The comments are pickled when `-Ykeep-comments` is set.

allanrenucci commented on compiler/test/dotty/tools/dotc/CommentPicklingTest.scala:
```diff
@@ -0,0 +1,178 @@
+import org.junit.Test
+import org.junit.Assert.{assertEquals, assertFalse, fail}
+
+class CommentPicklingTest extends ParallelTesting {
```
Why does this extends `ParallelTesting`?

allanrenucci commented on compiler/test/dotty/tools/dotc/CommentPicklingTest.scala:
```diff
@@ -0,0 +1,178 @@
+             maxDepth: Int = Int.MaxValue): Seq[Path] = {
+    val out = collection.mutable.ListBuffer.empty[Path]
+    val matcher = FileSystems.getDefault.getPathMatcher(pattern)
+    val visitor = new FileVisitor[Path] {
```
Something like?
```scala
val paths = Files.walk(base)
try paths.filter(matcher.matches).iterator().asScala.toList
finally paths.close()
```

allanrenucci commented on compiler/test/dotty/tools/dotc/CommentPicklingTest.scala:
```diff
@@ -0,0 +1,178 @@
+
+  private def getAll(base: Path,
+             pattern: String,
+             maxDepth: Int = Int.MaxValue): Seq[Path] = {
```
`maxDepth` is not used. Remove?

allanrenucci commented on compiler/test/dotty/tools/dotc/CommentPicklingTest.scala:
```diff
@@ -0,0 +1,178 @@
+    compileAndCheckComment(sources, "buzz".toTermName, Some("/** foo */"))
+  }
+
+  private def compileAndCheckComment(sources: Seq[String], treeName: Name, expectedComment: Option[String]): Unit = {
```
I would use `List` instead of `Seq` across the file. I don't think you gain anything by using `Seq`

allanrenucci replied:
Do we optimise the API for the most common use case where you want to check for the presence of a comment? `Option[String]` => `String`.

Also maybe add an overload for one source

allanrenucci commented on compiler/test/dotty/tools/dotc/CommentPicklingTest.scala:
```diff
@@ -0,0 +1,178 @@
+    }
+  }
+
+  private def findTreeNamed(name: Name)(trees: Seq[tpd.Tree], ctx: Context): Option[tpd.NameTree] = {
```
We have `find` on `Tree` in `tpd.TreeOps`

allanrenucci commented on compiler/test/dotty/tools/dotc/CommentPicklingTest.scala:
```diff
@@ -0,0 +1,178 @@
+package dotty.tools.dotc
```
Move to tasty package?

OlivierBlanvillain commented:
What do you think about reversing the flag, `-Ydrop-comments`? Python has great success with carrying comments everywhere!

Duhemm commented:
@OlivierBlanvillain I've tried switching `-Ykeep-comments` to `-Ydrop-comments`, but unfortunately this doesn't work with `-Ycheck` at the moment. There are several reasons for that:

- The `@usecase` annotations present in the docstrings introduce new symbols. `-Ycheck` will complain that the enclosing class has a symbol for which it doesn't find a declaration. This one is easy to solve by special-casing `-Ycheck` for those symbols (there must be a better solution, though...) https://github.com/lampepfl/dotty/blob/91de02962767cef1f6221e2bd565367aca1f3efc/compiler/src/dotty/tools/dotc/typer/Docstrings.scala#L26
- The symbols introduced by the `@usecase` annotations have names that are not valid on the JVM. This could probably be changed, but I imagine that Dottydoc extracts information from that name? This sounds like a hack to me. 
https://github.com/lampepfl/dotty/blob/91de02962767cef1f6221e2bd565367aca1f3efc/compiler/src/dotty/tools/dotc/typer/Docstrings.scala#L26
- I got other errors with `-Ycheck` in the body of the `@usecase` definition; I didn't debug them further.

I'd prefer to leave the switching of the flag for a later PR, when it's clearer to me what needs to be fixed in order to keep the comments by default.

nicolasstucki commented on compiler/src/dotty/tools/dotc/core/tasty/CommentPickler.scala:
```diff
@@ -0,0 +1,43 @@
+import java.nio.charset.Charset
+
+class CommentPickler(pickler: TastyPickler, addrOfTree: tpd.Tree => Option[Addr])(implicit ctx: Context) {
+  val buf = new TastyBuffer(5000)
```
`private[this]`

nicolasstucki commented on compiler/src/dotty/tools/dotc/core/tasty/TastyBuffer.scala:
```diff
@@ -48,6 +48,10 @@ class TastyBuffer(initialSize: Int) {
   // -- Output routines --------------------------------------------
 
+  /** Write a boolean value. */
+  def writeBoolean(b: Boolean): Unit =
```
I would write the byte explicitly `pickleComment` and read the byte in `CommentUnpickler`. This abstraction is only used once and hides the underlying representation. It would make it too easy to write two calls to `writeBoolean` instead of one to `writeByte` (with bitset representation).

nicolasstucki commented on compiler/src/dotty/tools/dotc/core/tasty/CommentPickler.scala:
```diff
@@ -0,0 +1,43 @@
+      buf.writeAddr(addr)
+      buf.writeNat(length)
+      buf.writeBytes(bytes, length)
+      buf.writeBoolean(cmt.isExpanded)
```
This seems to be missing from the documentation.

Duhemm replied:
What do you mean?

Duhemm replied:
I got it now

nicolasstucki commented on compiler/src/dotty/tools/dotc/core/tasty/TastyReader.scala:
```diff
@@ -41,6 +41,10 @@ class TastyReader(val bytes: Array[Byte], start: Int, end: Int, val base: Int =
     new TastyReader(bytes, index(start), index(end), base)
 
+  /** Read a boolean value. */
+  def readBoolean(): Boolean =
```
Same

nicolasstucki commented:
The `TastyPrinter` should print something for this section as is done for [positions](https://github.com/lampepfl/dotty/blob/master/compiler/src/dotty/tools/dotc/core/tasty/TastyPrinter.scala#L110-L121). You can test the printer with `dotc -print-tasty XYZ` for class `XYZ` in your classpath. This can be done in another PR.

Duhemm commented:
@nicolasstucki I've addressed your comments and modified the `TastyPrinter` to print the comments, too. Maybe it's too verbose to print the whole comments? I don't know how you usually use it, but it may be better to print only the first line of every comment?

nicolasstucki commented:
Maybe print the first `n` characters of the string and then add `...` and say on the side the actual length of the comment. `n` could be based on the current value of `-pagewidth`.

allanrenucci commented on compiler/src/dotty/tools/dotc/core/tasty/TreeUnpickler.scala:
```diff
@@ -820,6 +823,16 @@ class TreeUnpickler(reader: TastyReader,
         sym.registerIfChild(late = true)
+
+      if (ctx.mode.is(Mode.ReadComments)) {
+        for { docCtx <- ctx.docCtx
```
If the mode `ReadComments`, then I suppose we should both `docCtx` and `commentUnpickler` should not be empty

Duhemm replied:
The same goes for `posUnpicklerOpt`: we check if the mode is `ReadPositions` and whether `posUnpicklerOpt` is set.

The `TreeUnpickler` can be constructed with a certain configuration in mind (i.e. not defining `commentUnpicklerOpt`), and then used with a different configuration (i.e. `ctx.mode.is(Mode.ReadComments)`), which would lead to crashes if we just used `.get` on the `Option`.

Do we prefer crashing? Should we add a warning?

allanrenucci replied:
I really can't tell why the position unpickler is taken as an option.

I don't see a good reason to do the same. It only makes thing more complicated. I think the `TreeUnpickler` does not need an optional `CommentUnpickler`. I think we should just look at `Mode.ReadComments` to decide whether or not we unpickle comments. And in the case we do, we should assert that we have a `docCtx`.

Maybe @smarter or @nicolasstucki can explain why it was done like this for `PositionUnpickler`

smarter replied:
No idea, but what you said sounds reasonable

Duhemm replied:
It's actually valid for the `positionUnpickler` or `commentUnpickler` to be `None` if there are no `Positions` or `Comments` sections in the TASTY file from which we're unpickling.

allanrenucci commented on compiler/src/dotty/tools/dotc/Driver.scala:
```diff
@@ -40,10 +42,22 @@ class Driver extends DotClass {
+   * unpickle the doc comments.
+   */
+  protected def shouldAddDocContext(implicit ctx: Context): Boolean = {
+    !ctx.settings.YdropComments.value || ctx.mode.is(Mode.ReadComments)
```
Why do we have both a new mode and a compiler option?
Isn't `!ctx.settings.YdropComments <=> Mode.ReadComments`.

allanrenucci replied:
Fair enough

allanrenucci commented on compiler/src/dotty/tools/dotc/Driver.scala:
```diff
@@ -40,10 +42,22 @@ class Driver extends DotClass {
+   * to store doc comments unless `-Ydrop-comments` is set, or when TASTY is configured to
+   * unpickle the doc comments.
+   */
+  protected def shouldAddDocContext(implicit ctx: Context): Boolean = {
```
Likely or not, this is not currently the case. I would just inline it into `setup`. In the future, if there is a use case, one can always extract the wanted bits into a protected def

allanrenucci commented on compiler/src/dotty/tools/dotc/core/tasty/CommentPickler.scala:
```diff
@@ -0,0 +1,43 @@
+      ()
+  }
+
+  private class Traverser extends tpd.TreeTraverser {
```
Make the traverser take the `docCtx` and only create the traverser in `pickleComment` if it exists

allanrenucci commented on compiler/test/dotty/tools/dotc/core/tasty/CommentPicklingTest.scala:
```diff
@@ -0,0 +1,151 @@
+  }
+
+  private def findTreeNamed(name: Name)(trees: List[tpd.Tree], ctx: Context): Option[tpd.NameTree] = {
+    val acc = new tpd.TreeAccumulator[Option[tpd.NameTree]] {
```
why not use `TreeOps.find` from `tpd.scala`?

allanrenucci commented on compiler/test/dotty/tools/dotc/core/tasty/CommentPicklingTest.scala:
```diff
@@ -0,0 +1,151 @@
+      Files.createDirectories(out)
+
+      val options = compileOptions.and("-d", out.toAbsolutePath.toString).and(sourceFiles: _*)
+      val driver = new Driver
```
`dotty.tools.package.dotc.Main`?

allanrenucci commented on compiler/test/dotty/tools/dotc/core/tasty/CommentPicklingTest.scala:
```diff
@@ -0,0 +1,151 @@
+  }
+
+  private def compileAndUnpickle[T](sources: List[String])(fn: (List[tpd.Tree], Context) => T) = {
+    inTempDirectory { tmp =>
```
A nice to have (that shouldn't block this PR): a driver that compiles from virtual files. I think it would be fairly easy: override `doCompile` and call `compileSources(List[SourceFile])` instead of  `compile(List[String])`. Creating a virtual source file is straightforward: `new SourceFile("<fake name>", content.toCharArray)`

nicolasstucki commented on compiler/src/dotty/tools/dotc/core/quoted/TastyUnpickler.scala:
```diff
@@ -4,10 +4,10 @@ import dotty.tools.dotc.core.tasty._
 object TastyUnpickler {
-  class QuotedTreeSectionUnpickler(posUnpickler: Option[PositionUnpickler], splices: Seq[Any])
-    extends DottyUnpickler.TreeSectionUnpickler(posUnpickler) {
+  class QuotedTreeSectionUnpickler(posUnpickler: Option[PositionUnpickler], commentUnpickler: Option[CommentUnpickler], splices: Seq[Any])
```
I do not have a use case in mind. We should not add it.
